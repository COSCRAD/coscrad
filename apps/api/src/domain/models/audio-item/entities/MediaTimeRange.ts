import { NonEmptyString, NonNegativeFiniteNumber } from '@coscrad/data-types';
import { Maybe } from '../../../../lib/types/maybe';
import { NotFound } from '../../../../lib/types/not-found';
import { DTO } from '../../../../types/DTO';
import { isNullOrUndefined } from '../../../utilities/validation/is-null-or-undefined';
import BaseDomainModel from '../../BaseDomainModel';

// We can change this later
type MediaTimestamp = number;

// Update later
type CoscradText = string;

export class TranscriptItem<T extends string = string> extends BaseDomainModel {
    @NonNegativeFiniteNumber({
        label: 'in point',
        description: 'starting time stamp',
    })
    readonly inPoint: MediaTimestamp;

    @NonNegativeFiniteNumber({
        label: 'out point',
        description: 'ending time stamp',
    })
    readonly outPoint: MediaTimestamp;

    // TODO Abstract over different data types
    @NonEmptyString({
        label: 'text data',
        description: 'text for this time range',
    })
    readonly text: CoscradText;

    @NonEmptyString({
        label: 'label',
        description: 'the label for the current timestamped item',
    })
    readonly label: string;

    constructor(dto: DTO<TranscriptItem<T>>) {
        super();

        if (!dto) return;

        const { inPoint, outPoint, text: data, label } = dto;

        this.inPoint = inPoint;

        this.outPoint = outPoint;

        // TODO - clone if using a reference type for data
        if (data) this.text = data;

        this.label = label;
    }

    hasData(): boolean {
        return !isNullOrUndefined(this.text);
    }

    getData(): Maybe<CoscradText> {
        return this.hasData() ? this.text : NotFound;
    }

    setData<T extends CoscradText>(newData: T) {
        return this.clone<TranscriptItem<T>>({
            data: newData,
        } as unknown as TranscriptItem<T>);
    }
}
