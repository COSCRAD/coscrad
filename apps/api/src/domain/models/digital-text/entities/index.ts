export * from './digital-text-page.entity';
export * from './digital-text.entity';
export * from './types';
