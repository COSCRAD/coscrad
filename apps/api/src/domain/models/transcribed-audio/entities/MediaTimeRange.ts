import { Maybe } from 'apps/api/src/lib/types/maybe';
import { NotFound } from 'apps/api/src/lib/types/not-found';
import { DTO } from 'apps/api/src/types/DTO';
import { isNullOrUndefined } from '../../../utilities/validation/is-null-or-undefined';
import BaseDomainModel from '../../BaseDomainModel';

// We can change this later
type MediaTimestamp = number;

// Update later
type MediaData = string;

export class MediaTimeRange extends BaseDomainModel {
    readonly inPoint: MediaTimestamp;

    readonly outPoint: MediaTimestamp;

    readonly data?: MediaData;

    constructor({ inPoint, outPoint, data }: DTO<MediaTimeRange>) {
        super();

        this.inPoint = inPoint;

        this.outPoint = outPoint;

        // TODO - clone if using a reference type for data
        if (data) this.data = data;
    }

    hasData(): boolean {
        return !isNullOrUndefined(this.data);
    }

    getData(): Maybe<MediaData> {
        return this.hasData() ? this.data : NotFound;
    }

    setData(newData: MediaData) {
        return this.clone<MediaTimeRange>({
            data: newData,
        });
    }
}
