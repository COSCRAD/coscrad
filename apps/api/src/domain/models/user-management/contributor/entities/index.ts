export * from './coscrad-contributor.entity';
export * from './coscrad-contributor.module';
