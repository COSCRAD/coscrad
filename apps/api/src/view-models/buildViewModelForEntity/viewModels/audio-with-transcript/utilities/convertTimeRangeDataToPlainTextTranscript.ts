import { MediaTimeRange } from 'apps/api/src/domain/models/audio-with-transcript/entities/MediaTimeRange';

export default (timeranges: MediaTimeRange[]): string =>
    timeranges.reduce(
        (accumulatedPlainText, { inPoint, outPoint, data }) =>
            accumulatedPlainText.concat(`\n[${inPoint}] ${data} [${outPoint}]`),
        ''
    );
