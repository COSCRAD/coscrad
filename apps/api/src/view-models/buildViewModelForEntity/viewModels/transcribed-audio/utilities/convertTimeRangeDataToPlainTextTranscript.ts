import { MediaTimeRange } from 'apps/api/src/domain/models/transcribed-audio/entities/MediaTimeRange';

export default (timeranges: MediaTimeRange[]): string =>
    timeranges.reduce(
        (accumulatedPlainText, { inPoint, outPoint, data }) =>
            accumulatedPlainText.concat(`\n[${inPoint}] ${data} [${outPoint}]`),
        ''
    );
