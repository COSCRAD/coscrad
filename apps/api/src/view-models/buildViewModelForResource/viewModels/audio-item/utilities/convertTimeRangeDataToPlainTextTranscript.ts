import { TranscriptItem } from '../../../../../domain/models/audio-item/entities/MediaTimeRange';

export default (timeranges: TranscriptItem[]): string =>
    timeranges.reduce(
        (accumulatedPlainText, { inPoint, outPoint, text: data }) =>
            accumulatedPlainText.concat(`\n[${inPoint}] ${data} [${outPoint}]`),
        ''
    );
