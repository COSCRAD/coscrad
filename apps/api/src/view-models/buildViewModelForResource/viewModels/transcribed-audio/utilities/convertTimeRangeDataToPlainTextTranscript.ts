import { MediaTimeRange } from '../../../../../domain/models/transcribed-audio/entities/MediaTimeRange';

export default (timeranges: MediaTimeRange[]): string =>
    timeranges.reduce(
        (accumulatedPlainText, { inPoint, outPoint, data }) =>
            accumulatedPlainText.concat(`\n[${inPoint}] ${data} [${outPoint}]`),
        ''
    );
