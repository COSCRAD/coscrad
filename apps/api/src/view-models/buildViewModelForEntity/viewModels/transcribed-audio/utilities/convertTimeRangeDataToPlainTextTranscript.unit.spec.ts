import { MediaTimeRange } from 'apps/api/src/domain/models/transcribed-audio/entities/MediaTimeRange';
import convertTimeRangeDataToPlainTextTranscript from './convertTimeRangeDataToPlainTextTranscript';

// TODO- add this to the test data and import it here!
const testTranscript: MediaTimeRange[] = [
    'Once upon a time, not long ago, there lived three dogs.',
    'Each of these dogs had a collar of a different color.',
    'One of these dogs did not obey his caller.',
    "This dog's collar was of a reddish color.",
    'One day the dog whose collar was a reddish color got into the collards',
    'to the spite of his caller.',
]
    .map((text, index) => ({
        inPoint: index * 500,
        outPoint: index * 500 + 400,
        data: text,
    }))
    .map((dto) => new MediaTimeRange(dto));

describe('convertTimeRangeDataToPlainTextTranscript', () => {
    describe('when the input is an empty array', () => {
        it('should return an empty string', () => {
            const result = convertTimeRangeDataToPlainTextTranscript([]);

            expect(result).toBe('');
        });
    });

    describe('when given data for a sample transcript', () => {
        it('should produce the expected plain text', () => {
            const result = convertTimeRangeDataToPlainTextTranscript(testTranscript);

            expect(result).toMatchSnapshot();
        });
    });
});
