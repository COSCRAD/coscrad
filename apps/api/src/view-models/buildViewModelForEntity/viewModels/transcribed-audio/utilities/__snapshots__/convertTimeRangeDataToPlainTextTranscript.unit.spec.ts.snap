// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertTimeRangeDataToPlainTextTranscript when given data for a sample transcript should produce the expected plain text 1`] = `
"
[0] Once upon a time, not long ago, there lived three dogs. [400]
[500] Each of these dogs had a collar of a different color. [900]
[1000] One of these dogs did not obey his caller. [1400]
[1500] This dog's collar was of a reddish color. [1900]
[2000] One day the dog whose collar was a reddish color got into the collards [2400]
[2500] to the spite of his caller. [2900]"
`;
